src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -235,64 +299,72 @@ export class SavedObjectsService
-      },
-    });
-
-    return {
-      getScopedClient: this.clientProvider.getClient.bind(this.clientProvider),
-      setClientFactory: this.clientProvider.setClientFactory.bind(this.clientProvider),
-      addClientWrapper: this.clientProvider.addClientWrapperFactory.bind(this.clientProvider),
-      createInternalRepository: (extraTypes?: string[]) =>
-        createSORepository(adminClient.callAsInternalUser, extraTypes),
-      createScopedRepository: (req: KibanaRequest, extraTypes?: string[]) =>
-        createSORepository(adminClient.asScoped(req).callAsCurrentUser, extraTypes),
-    };
+    await migrator.runMigrations(skipMigrations);
   }
 
   public async start(core: SavedObjectsStartDeps): Promise<InternalSavedObjectsServiceStart> {
-    if (!this.clientProvider) {
+    if (!this.setupDeps) {
       throw new Error('#setup() needs to be run first');
     }
+    if (!this.createRepository) {
+      throw new Error('#runMigration() needs to be run first');
+    }
 
     this.logger.debug('Starting SavedObjects service');
 
-    /**
-     * Note: We want to ensure that migrations have completed before
-     * continuing with further Core startup steps that might use SavedObjects
-     * such as running the legacy server, legacy plugins and allowing incoming
-     * HTTP requests.
-     *
-     * However, our build system optimize step and some tests depend on the
-     * HTTP server running without an Elasticsearch server being available.
-     * So, when the `migrations.skip` is true, we skip migrations altogether.
-     */
-    const cliArgs = this.coreContext.env.cliArgs;
-    const savedObjectsConfig = await this.coreContext.configService
-      .atPath<SavedObjectsConfigType>('migrations')
-      .pipe(first())
-      .toPromise();
-    const skipMigrations = cliArgs.optimize || savedObjectsConfig.skip;
-    await this.migrator!.runMigrations(skipMigrations);
+    const adminClient = this.setupDeps!.elasticsearch.adminClient;
+
+    const repositoryFactory: SavedObjectsRepositoryFactory = {
+      createInternalRepository: (extraTypes?: string[]) =>
+        this.createRepository!(adminClient.callAsInternalUser, extraTypes),
+      createScopedRepository: (req: KibanaRequest, extraTypes?: string[]) =>
+        this.createRepository!(adminClient.asScoped(req).callAsCurrentUser, extraTypes),
+    };
+
+    const clientProvider = new SavedObjectsClientProvider<KibanaRequest>({
+      defaultClientFactory({ request }) {
+        const repository = repositoryFactory.createScopedRepository(request);
+        return new SavedObjectsClient(repository);
+      },
+    });
+    if (this.clientFactoryProvider) {
+      const clientFactory = this.clientFactoryProvider(repositoryFactory);
+      clientProvider.setClientFactory(clientFactory);
+    }
+    this.clientFactoryWrappers.forEach(({ id, factory, priority }) => {
+      clientProvider.addClientWrapperFactory(priority, id, factory);
+    });
```

Comment:
Now that the clientProvider is created after everything has been registered to the service, we should probably use a simple constructor instead, 
```
const clientProvider = new SavedObjectsClientProvider<KibanaRequest>({
   clientFactory,
   clientWrappers,
})
```

Reply:
I tried to do it, but TIL is that the clientProvider is exposed in the internal API to be directly used from the legacy SO service in `src/legacy/server/saved_objects/saved_objects_mixin.js`, so changing the class is not trivial. This should wait until legacy has been removed.